man/man1/rastrip.1
.\" Copyright (c) 2000-2008 QoSient, LLC
.\" All rights reserved.
.\" 
.\" This program is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 2, or (at your option)
.\" any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program; if not, write to the Free Software
.\" Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.TH RASTRIP 1 "07 November 2000"
.SH NAME
\fBrastrip\fP \- strip \fBargus(8)\fP data file.
.SH COPYRIGHT
Copyright (c) 2000-2008 QoSient. All rights reserved.
.SH SYNOPSIS
.B rastrip
[\fB\-M\fP [replace] [+|-]\fIdsr\fP [-M ...]]
[\fBraoptions\fP]
.SH DESCRIPTION
.IX  "rastrip command"  ""  "\fLrastrip\fP \(em argus data"
.LP
.B Rastrip
reads
.BR argus
data from an \fIargus-data\fP source, strips the records based on
the criteria specified on the command line, and outputs a valid
\fIargus-stream\fP. This is useful to reduce the size of argus data files.
Rastrip always removes argus management transactions, thus having
the same effect as a 'not man' filter expression. 
.SH OPTIONS
Rastrip, like all ra based clients, supports a number of
\fBra options\fP including filtering of input argus
records through a terminating filter expression.
See \fBra(1)\fP for a complete description of \fBra options\fP.
\fBrastrip(1)\fP specific options are:
.PP
.PD 0
.TP 4 4
.B \-M [replace] [+|-]dsr
Strip specified dsr (data structure record?).

Supported dsrs are:
.PP
.RS
.TP
.B flow
flow key data (proto, saddr, sport, dir, daddr, dport)
.TP
.B time
time stamp fields (stime, ltime).
.TP
.B metric
basic ([s|d]bytes, [s|d]pkts, [s|d]rate, [s|d]load)
.TP
.B agr
aggregation stats (trans, avgdur, mindur, maxdur, stdev).
.TP
.B net
network objects (tcp, esp, rtp, icmp data).
.TP
.B vlan
VLAN tag data 
.TP
.B mpls
MPLS label data
.TP
.B jitter
Jitter data ([s|d]jit, [s|d]intpkt)
.TP
.B ipattr
IP attributes ([s|d]ipid, [s|d]tos, [s|d]dsb, [s|d]ttl)
.TP
.B suser
src user captured data bytes (suser)
.TP
.B duser
dst captured user data bytes (duser)
.TP
.B mac
MAC addresses (smac, dmac)
.TP
.B icmp
ICMP specific data (icmpmap, inode)
.TP
.B encaps
Flow encapsulation type indications
.PD
.RE

If no dsrs are specified, Rastrip removes the following default set of dsrs: encaps, agr, vlan, mpls, mac, icmp, ipattr, jitter, suser, duser

.SH INVOCATION
A sample invocation of \fBrastrip(1)\fP.  This call reads \fBargus(8)\fP data
from \fBinputfile\fP and strips the default dsr set but keeps MAC addresses and writes the result
to \fBoutputfile\fP:

\fBrastrip -M +mac -r inputfile -w outputfile\fP

This call removes only user captured data and timings and writes the result to stdout:

\fBrastrip -M -suser -M -duser -M -time -r inputfile\fP

.SH SEE ALSO
.BR ra(1),
.BR rarc(5),
.BR argus(8),
.SH FILES

.SH AUTHORS
.nf
Carter Bullard (carter@qosient.com).
.fi
.SH BUGS


ChangeLog
Fri Jul 31 11:47:29 EDT 2009
Fix for 2.0 record conversion.  rtcp dsr inserted with len of 1 and
ArgusNtoH was corrupting other DSRs trying to little-endian the field.

Wed Jul 29 13:46:38 EDT 2009
Wow, "-S 1.2.3.4" didn't work, so fixed that.

Wed Jul 22 00:46:11 EDT 2009
Fix for trans processing in rahisto().
Modification of 'autoid' support in rasqlinsert().

Thu Jul  9 15:35:05 EDT 2009
Fix for rasplit() bug where the $srcid directive is ignored in some
situations. Not good.

Wed Jul  8 15:49:13 EDT 2009
Fix for TCP synAck flows.  Extend appbyte check in case argus wasn't 
configured to generate appbytes.
Fix for ratop ^L key.

Fri Jun 19 12:20:24 EDT 2009
Fix for ASN for inode IP address little endian converstion problem.

Mon Jun 15 09:12:25 EDT 2009
Major fix for processing IPv6 ICMP data.
Major fix for processing User data DSR.
Fix to not print field delimiter when  printing fields.

Thu May 21 13:11:59 EDT 2009
New fix for rasplit when spliting on anything bigger than hours.
New faster strategy for dealing with daylight savings time issues.
Updates for rabins for new spliting strategies.

Fri May 15 13:26:09 EDT 2009
Mods to fix RA_FIELD_WIDTH behavior.

Thu May  7 12:47:06 EDT 2009
Massive changes for Geolocation.
AS number reporting (change as from short to int).
Improved DNS reporting formats so field number is constant.
Fixed merging record problem.

Mon Apr 13 17:55:47 EDT 2009
Modified RA_FIELD_WIDTH behavior with using '-c' on the command line.
Added RA_FIELD_WIDTH to rarc.

Fixed problem with RaPrintIndex and rasqlinsert.
Modified rasqlinsert() minimum field width to be 256, so that addresses
will always be big enough.

Modified RA_TIME_FORMAT string parsing ala Dan Klien, to get non time
oriented characters and escaped sequences parsed correctly.

Mon Apr  6 11:37:36 EDT 2009
Fixed v2 -> v3 conversion.  ArgusNetworkStruct len was a few bytes too long.
Fixed mode parsing for rasplit().
Fix for rasplit() finding filename on split 'd','w','m','y'.  Algorithm
chose GMT not local timezone.

Fri Apr  3 12:53:15 EDT 2009
Fixes for rasqlinsert() to better deal with strftime() based table names.
Fixed ratree() to work in argus-3.0.

Mon Mar 30 16:34:24 EDT 2009
Added table date specifications for rasqlinsert().
Fix for memory problem when opening lots of files or directories.
Added RA_TIMEOUT_INTERVAL to rarc file.
Fixed short DSR parsing.

Fri Mar 27 09:41:05 EDT 2009
Fix for tcprtt propagation during merge when first record
is a ARGUS_TCP_INIT type DSR.

Fix for little-endian treatment of Authoritative response
in DNS printing routines of radump().

Wed Mar 25 16:09:53 EDT 2009
Mods to complete library to fix issues with -M rmon option.
Problem caused duplicate records generated by "-M rmon"
to be corrected before being aggregated.  Big problem,
basically just doubling metrics, but not flow ids.

Mon Mar 23 14:51:21 EDT 2009
Mod to ArgusClientInit in rabins() to deal with calling without
a -B option to specify the range.  Seems to work now.

Fri Mar 13 14:53:46 EDT 2009
Minor changes for raservices and rauserdata.
May have broken "finding" possible protocol types, need
to work on this this  weekend.

Thu Mar 12 11:40:14 EDT 2009
Removed -lz from aclocal.m4 an configure.in.  Will need
to add it back soon.

Wed Mar 11 20:30:17 EDT 2009
Fixed record searching when the read buffer boundary is
too short for searching for new data record.  Reordered
logic to do test, when enough bytes are available, and
to break way out if not enough bytes.

Mon Mar  9 12:45:03 EDT 2009
Fixed aggregation rules for FRAGMENT reporting in flow.
argus changes its reporting of FRAGMENTS scope to the single
flow record, and that revealed that merging was messing
up the IP attributes flags.  

Fri Mar  6 13:38:36 EST 2009
Added record searching for argus-2.x data sets when out of sync.
fixed early file closing for ra*.

Mon Mar  2 14:28:02 EST 2009
Release candidate #1 for clients-3.0.2.  This has all the support
for next wave argus DSRs (histograms and events), and many bug
fixes from the mailing list and beyond.

Mods to TCP direction corrections, to be more strict on what
we'll reverse.

Fixes for backward compatiblity for argus-2.x data streams.

Added rauserdata and rasservices to distribution to assist in user
data analysis.

Wed Feb 25 16:01:36 EST 2009
Complete integration of gargoyle client code back to argus,
without IB and control plane flow monitoring.  Major
rewrite of rabins(), ratop() and racluster() for ARMA
functions.  Completely differnt compiler filtering strategies
to allow for DSR specific filtering.

Addition of mysql insertion and query programs, but no NTAIS
project support.

Elimination of ragrep(), by merging grep capabilities into all
ra* programs.

Tue Feb 24 10:07:24 EST 2009
Major mods to fix graphing problems exposed by ArgusGraph and Topology.
Added sub-second bins, and better reliablity.

Tue Oct 28 23:32:29 EDT 2008
Major modifications for introducing new linear histogram
reporting for interpacket arrival times.

Mon Oct 27 10:44:42 EDT 2008
Major modifications for restructuring library strategies to accomodate
new client architectures.

Fri Sep  5 13:27:01 EDT 2008
Fixes for rasql, rasqlinsert, all is well.
Terminatd xml ArgusDataStream in rasql.

Fri Aug  1 10:39:35 EDT 2008
Massive changes for all things.
MySQL programs now working well.

Tue Apr  8 14:55:09 EDT 2008
Fixes for time filter problem and addition of UTC input.
Fixes for radump() to parse user buffer.

Sat Mar 15 17:14:31 EDT 2008
Fixes for printing when srcid is not an IP address.

Thu Feb 28 19:38:40 EST 2008
Fixes for rastream() and to get the right readline
on the mac.

Wed Feb 13 00:17:15 EST 2008
Added support for Multipath

Mon Feb 11 22:45:40 EST 2008
Mods for putting listen() down on all address families.
Mod for Makefiles to remove .PHONY for OpenBSD

Tue Jan 29 13:36:49 EST 2008
Fixes for new flex 2.5.34 issues

Mon Jan 21 20:03:26 EST 2008
Fixes for default unidirectional flow record matching

clients/rastrip.c
/*
 * Argus Software
 * Copyright (c) 2000-2009 QoSient, LLC
 * All rights reserved.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.

 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 * 
 * $Id: //depot/argus/clients/clients/rastrip.c#22 $
 * $DateTime: 2009/05/15 12:40:06 $
 * $Change: 1732 $
 */

/*
 * rastrip.c  - remove fields from argus records.
 *       
 * written by Carter Bullard
 * QoSient, LLC
 *
 */


#if defined(CYGWIN)
#define USE_IPV6
#endif

#include <unistd.h>
#include <stdlib.h>

#include <signal.h>
#include <ctype.h>

#include <compat.h>

#include <rabins.h>
#include <argus_util.h>
#include <argus_client.h>
#include <argus_main.h>
#include <argus_filter.h>

#define ARGUS_ADD_OPTION		1
#define ARGUS_SUB_OPTION		2


extern char *ArgusDSRKeyWords[ARGUSMAXDSRTYPE];
int ArgusDSRFields[ARGUSMAXDSRTYPE];

/*
#define ARGUS_TRANSPORT_INDEX            0
#define ARGUS_FLOW_INDEX                 1
#define ARGUS_TIME_INDEX                 2
#define ARGUS_METRIC_INDEX               3
#define ARGUS_AGR_INDEX                  4
#define ARGUS_NETWORK_INDEX              5 
#define ARGUS_VLAN_INDEX                 6 
#define ARGUS_MPLS_INDEX                 7
#define ARGUS_JITTER_INDEX               8
#define ARGUS_IPATTR_INDEX               9
#define ARGUS_PSIZE_INDEX               10
#define ARGUS_SRCUSERDATA_INDEX         11
#define ARGUS_DSTUSERDATA_INDEX         12
#define ARGUS_MAC_INDEX                 13 
#define ARGUS_ICMP_INDEX                14
#define ARGUS_ENCAPS_INDEX              15
#define ARGUS_TIME_ADJ_INDEX            16
#define ARGUS_COR_INDEX                 17
#define ARGUS_COCODE_INDEX              18
                                                                                                                                                          
char *ArgusDSRKeyWords[ARGUSMAXDSRTYPE] = {
   "trans",
   "flow",
   "time",
   "metric",
   "agr",
   "net",
   "vlan",
   "mpls",
   "jitter",
   "ipattr",
   "psize",
   "suser",
   "duser",
   "mac",
   "icmp",
   "encaps",
   "tadj",
   "cor",
   "cocode",
};
*/

int ArgusReplaceMode = 0;
int ArgusStripDown = 0;
int ArgusFirstMOptionField = 1;

void ArgusProcessOptions(struct ArgusModeStruct *);

void
ArgusClientInit (struct ArgusParserStruct *parser)
{
   struct ArgusModeStruct *mode = NULL;

   parser->RaWriteOut = 0;

   if (!(parser->RaInitialized)) {
      (void) signal (SIGHUP,  (void (*)(int)) RaParseComplete);
      (void) signal (SIGTERM, (void (*)(int)) RaParseComplete);
      (void) signal (SIGQUIT, (void (*)(int)) RaParseComplete);
      (void) signal (SIGINT,  (void (*)(int)) RaParseComplete);

      parser->RaInitialized++;

      if ((mode = parser->ArgusModeList) != NULL) {
         while (mode) {
            if (!(strcmp ("replace", mode->mode))) {
               ArgusReplaceMode++;
               if ((parser->ArgusWfileList != NULL) && (!(ArgusListEmpty(parser->ArgusWfileList)))) {
                  ArgusLog (LOG_ERR, "replace mode and -w option are incompatible\n");
               }
               break;
            }
            mode = mode->nxt;
         }
      }

      if ((mode = parser->ArgusModeList) != NULL)
         ArgusProcessOptions(mode);
      else {
         bzero ((char *)ArgusDSRFields, sizeof(ArgusDSRFields));
         ArgusDSRFields[ARGUS_TIME_INDEX] = 1;
         ArgusDSRFields[ARGUS_FLOW_INDEX] = 1;
         ArgusDSRFields[ARGUS_METRIC_INDEX] = 1;
         ArgusDSRFields[ARGUS_NETWORK_INDEX] = 1;
      }

      if (!(ArgusStripDown))
         ArgusStripDown++;
   }
}

void RaArgusInputComplete (struct ArgusInput *input) { return; }

void
RaParseComplete (int sig)
{
   if (sig >= 0) {
      if (!ArgusParser->RaParseCompleting++) {
#ifdef ARGUSDEBUG
         ArgusDebug (2, "RaParseComplete(caught signal %d)\n", sig);
#endif
         switch (sig) {
            case SIGHUP:
            case SIGINT:
            case SIGTERM:
            case SIGQUIT: {
               struct ArgusWfileStruct *wfile = NULL;

               if (ArgusParser->ArgusWfileList != NULL) {
                  struct ArgusListObjectStruct *lobj = NULL;
                  int i, count = ArgusParser->ArgusWfileList->count;

                  if ((lobj = ArgusParser->ArgusWfileList->start) != NULL) {
                     for (i = 0; i < count; i++) {
                        if ((wfile = (struct ArgusWfileStruct *) lobj) != NULL) {
                           if (wfile->fd != NULL) {
#ifdef ARGUSDEBUG
                              ArgusDebug (2, "RaParseComplete: closing %s\n", wfile->filename);
#endif
                              fflush (wfile->fd);
                              fclose (wfile->fd);
                              wfile->fd = NULL;
                           }
                        }
                        lobj = lobj->nxt;
                     }
                  }
               }
               exit(0);
               break;
            }
         }
      }
   }
}


void
ArgusClientTimeout ()
{
#ifdef ARGUSDEBUG
   ArgusDebug (6, "ArgusClientTimeout()\n");
#endif
}

void
parse_arg (int argc, char**argv)
{}

void
usage ()
{
   extern char version[];

   fprintf (stderr, "Rastrip Version %s\n", version);
   fprintf (stderr, "usage: %s [-M [modes] [+|-]dsr [dsr ...]] [ra-options]\n\n", ArgusParser->ArgusProgramName);

   fprintf (stderr, "options: -M replace      strip dsrs and overwrite current file.\n");
   fprintf (stderr, "            [+|-] dsrs   [add|subtract] dsrs from records.\n");
   fprintf (stderr, "               dsrs:     stime, ltime, count, dur, avgdur,\n");
   fprintf (stderr, "                         srcid, ind, mac, dir, jitter, status, user,\n");
   fprintf (stderr, "                         win, trans, seq, vlan, vid, vpri, mpls.\n");

   exit(1);
}


void
RaProcessRecord (struct ArgusParserStruct *parser, struct ArgusRecordStruct *argus)
{
   static char buf[MAXSTRLEN];
   int x;

   switch (argus->hdr.type & 0xF0) {
      case ARGUS_MAR:
      case ARGUS_EVENT: {
         break;
      }

      case ARGUS_NETFLOW:
      case ARGUS_FAR: {
         for (x = 0; x < ARGUSMAXDSRTYPE; x++) {
            if (!(ArgusDSRFields[x])) {
               argus->dsrs[x] = NULL;
            }
         }
         
         if (parser->ArgusWfileList != NULL) {
            struct ArgusWfileStruct *wfile = NULL;
            struct ArgusListObjectStruct *lobj = NULL;
            int i, count = parser->ArgusWfileList->count;

            if ((lobj = parser->ArgusWfileList->start) != NULL) {
               for (i = 0; i < count; i++) {
                  if ((wfile = (struct ArgusWfileStruct *) lobj) != NULL) {
                     int pass = 1;
                     if (wfile->filterstr) {
                        struct nff_insn *wfcode = wfile->filter.bf_insns;
                        pass = ArgusFilterRecord (wfcode, argus);
                     }

                     if (pass != 0) {
                        if ((parser->exceptfile == NULL) || strcmp(wfile->filename, parser->exceptfile)) {
                           struct ArgusRecord *argusrec = NULL;
                           static char sbuf[0x10000];
                           if ((argusrec = ArgusGenerateRecord (argus, 0L, sbuf)) != NULL) {
#ifdef _LITTLE_ENDIAN
                              ArgusHtoN(argusrec);
#endif
                              ArgusWriteNewLogfile (parser, argus->input, wfile, argusrec);
                           }
                        }
                     }
                  }

                  lobj = lobj->nxt;
               }
            }

         } else {
            if (!parser->qflag) {
               if (parser->Lflag) {
                  if (parser->RaLabel == NULL)
                     parser->RaLabel = ArgusGenerateLabel(parser, argus);
       
                  if (!(parser->RaLabelCounter++ % parser->Lflag))
                     printf ("%s\n", parser->RaLabel);
       
                  if (parser->Lflag < 0)
                     parser->Lflag = 0;
               }

               *(int *)&buf = 0;
               ArgusPrintRecord(parser, buf, argus, MAXSTRLEN);
               fprintf (stdout, "%s\n", buf);
            }
         }
         break;
      }
   }
}

int RaSendArgusRecord(struct ArgusRecordStruct *argus) {return 0;}

void ArgusWindowClose(void);

void ArgusWindowClose(void) { 
#ifdef ARGUSDEBUG
   ArgusDebug (6, "ArgusWindowClose () returning\n"); 
#endif
}


void
ArgusProcessOptions(struct ArgusModeStruct *mode)
{
   int x, RaOptionOperation, setValue = 0;
   char *ptr = NULL;
   char *endptr;

   if (mode != NULL) {
      while (mode) {
         if ((strcmp ("replace", mode->mode))) {
            if (isdigit((int)*mode->mode)) {
                ArgusStripDown = strtol(mode->mode, &endptr, 10);
                if (mode->mode == endptr)
                   usage();

            } else {
               if (*mode->mode == '-') {
                  if (ArgusFirstMOptionField) {
                     for (x = 0; x < ARGUSMAXDSRTYPE; x++)
                        ArgusDSRFields[x] = 1;
                     ArgusFirstMOptionField = 0;
                  }
                  ptr = mode->mode + 1;
                  RaOptionOperation = ARGUS_SUB_OPTION;
               } else 
               if (*mode->mode == '+') {
                  if (ArgusFirstMOptionField) {
                     bzero ((char *)ArgusDSRFields, sizeof(ArgusDSRFields));
                     ArgusDSRFields[ARGUS_TIME_INDEX] = 1;
                     ArgusDSRFields[ARGUS_FLOW_INDEX] = 1;
                     ArgusDSRFields[ARGUS_METRIC_INDEX] = 1;
                     ArgusDSRFields[ARGUS_NETWORK_INDEX] = 1;
                  }
                  ptr = mode->mode + 1;
                  RaOptionOperation = ARGUS_ADD_OPTION;
               } else {
                  if (ArgusFirstMOptionField) {
                     bzero ((char *) ArgusDSRFields, sizeof(ArgusDSRFields));
                     ArgusFirstMOptionField = 0;
                  }
                  ptr = mode->mode;
                  RaOptionOperation = ARGUS_ADD_OPTION;
               }

               setValue = (RaOptionOperation == ARGUS_ADD_OPTION) ? 1 : 0;

               for (x = 0; x < ARGUSMAXDSRTYPE; x++) {
                  if (strlen(ArgusDSRKeyWords[x])) {
                     if (!strncmp (ArgusDSRKeyWords[x], ptr, strlen(ArgusDSRKeyWords[x]))) {
                        ArgusDSRFields[x] = setValue;
                        break;
                     }
                  }
               }
            }
         }

         mode = mode->nxt;
      }
   }
}
